python-package/lightgbm/sklearn.py

```diff
@@ -897,7 +897,7 @@ def n_features_(self) -> int:
         return self._n_features
 
     @property
-    def n_features_in_(self) -> int:
+    def n_features_in_(self) -> Optional[int]:
```

Comment:
I think you're missing some important context around how these `n_features_in_` and `best_iteration_` attributes work.

As you can see in the methods, before returning a value the function `self.__sklearn_is_fitted__()` is called. That method returns `True` if `self.fitted_` exists and is `True` (see #4636).

https://github.com/microsoft/LightGBM/blob/29796eee6eabb79e161880976ec02fcd95dfcc4a/python-package/lightgbm/sklearn.py#L548-L549

If `self.__sklearn_is_fitted__()` returns `True`, the attributes referenced by `self.n_features_in_` and `self.best_iteration_` should not be `None`. Fitting a model should set all of those attributes.

But `mypy` can't possibly know about those specifics of how `lightgbm` is expected to work and be used, and it's rightly complaining that "hey, you initialized `self.best_iteration` to `None` and now you're returning it...therefore it's possible for that value to be `None`".

I don't want to change these type hints to `Optional[None]` because then any other code relying on these attributes (including code in projects downstream of `lightgbm` like `optuna`) will have to face this awkwardness and either use `# type: ignore` or add defensive code checking like `if model.best_iteration is not None: ...`.

I tried one approach to work around these back in 2021 (https://github.com/microsoft/LightGBM/pull/4837#discussion_r759757551) that we decided to not adopt, and looking back now I agree that adding assertions or other new complexity might not be the best way to handle this.

I think a better approach would be to initialize those attributes to values of the expected return type, e.g. to initialize `self.best_score_` to `-1.0`.

But I think that needs to be done by someone who's more familiar with LightGBM, like one of the maintainers here, because it's important to, for example, check for cases where those internal attributes are being compared to `None`.

Ok so! After all that.... please just remove these changes to `best_iteration_` and `n_features_in` from this PR. I will address them in a separate PR.

Reply:
I just put up #5718 with a proposed fix.